### ISSUE TITLE
torch.compile: wrong answer on recursive function with slicing

### ISSUE THREAD

**Issue by albertfgu** [pytorch/pytorch#99833]
### 🐛 Describe the bug

Below is a simple function that takes a tensor and returns it unchanged. When `torch.compile`d, it returns wrong answers for length >= 4. Note that the wrong answers are different every call (probably indicating some sort of memory corruption).

(Context: this started from a port of https://jax.readthedocs.io/en/latest/_modules/jax/_src/lax/control_flow/loops.html#associative_scan,
which ran into this bug and I substantially simplified it to a minimal example)

```python
import torch

def recursive_identity(x):
    length = x.shape[-1]
    if length <= 1: return x

    x_0 = x[..., :length//2]
    x_1 = x[..., length//2:]

    y_1 = recursive_identity(x_1)
    y_0 = x_0

    y = torch.cat([y_0, y_1], dim=-1)
    return y

def test(L):
    a = 1 + torch.arange(L)

    csum = recursive_identity(a)
    print(csum)

    # Compile function
    recursive_identity_opt = torch.compile(recursive_identity)
    csum = recursive_identity_opt(a)
    print(csum)


if __name__ == "__main__":
    test(4)
```

-------------------
Output:
```
❯ TORCHDYNAMO_REPRO_AFTER="aot" python -m src.models.functional.scan_bug
tensor([1, 2, 3, 4])
tensor([              1,               2,               3, 140606498878752])

❯ TORCHDYNAMO_REPRO_AFTER="aot" python -m src.models.functional.scan_bug
tensor([1, 2, 3, 4])
tensor([              1,               2,               3, 139705624228976])
```

### Error logs

N/A

### Minified repro

`TORCHDYNAMO_REPRO_AFTER="aot"` gives the output up top.

`dynamo` crashes:
```
❯ TORCHDYNAMO_REPRO_AFTER="dynamo" python -m scan_bug
tensor([1, 2, 3, 4])
zsh: segmentation fault  TORCHDYNAMO_REPRO_AFTER="dynamo" python -m src.models.functional.scan_bug
```

### Versions

This script does not exist (404 Error) by the way - it got updated from `master` to `main`

```
❯ python collect_env.py
Collecting environment information...
PyTorch version: 2.0.0
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 10 (buster) (x86_64)
GCC version: (Debian 8.3.0-6) 8.3.0
Clang version: Could not collect
CMake version: version 3.26.3
Libc version: glibc-2.28

Python version: 3.10.10 (main, Mar 21 2023, 18:45:11) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.19.0-23-cloud-amd64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: 11.3.109
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-40GB
Nvidia driver version: 510.47.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
Address sizes:       46 bits physical, 48 bits virtual
CPU(s):              12
On-line CPU(s) list: 0-11
Thread(s) per core:  2
Core(s) per socket:  6
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) CPU @ 2.20GHz
Stepping:            7
CPU MHz:             2200.164
BogoMIPS:            4400.32
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            39424K
NUMA node0 CPU(s):   0-11
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat avx512_vnni md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] pytorch-lightning==1.9.3
[pip3] torch==2.0.0
[pip3] torchaudio==2.0.0
[pip3] torchdata==0.6.0
[pip3] torchmetrics==0.11.4
[pip3] torchtext==0.15.1
[pip3] torchvision==0.15.0
[pip3] triton==2.0.0
[conda] blas                      1.0                         mkl
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0           py310h7f8727e_0
[conda] mkl_fft                   1.3.1           py310hd6ae3a3_0
[conda] mkl_random                1.2.2           py310h00e6091_0
[conda] numpy                     1.23.5          py310hd5efca6_0
[conda] numpy-base                1.23.5          py310h8e6c178_0
[conda] pytorch                   2.0.0           py3.10_cuda11.7_cudnn8.5.0_0    pytorch
[conda] pytorch-cuda              11.7                 h778d358_3    pytorch
[conda] pytorch-lightning         1.9.3                    pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.0.0               py310_cu117    pytorch
[conda] torchdata                 0.6.0                    pypi_0    pypi
[conda] torchmetrics              0.11.4                   pypi_0    pypi
[conda] torchtext                 0.15.1                   pypi_0    pypi
[conda] torchtriton               2.0.0                     py310    pytorch
[conda] torchvision               0.15.0              py310_cu117    pytorch
```

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

**Comment by yanboliang** [comment]
@albertfgu I can't reproduce w/ your repro again main branch, can you verify it?

**Comment by Chillee** [comment]
Also cc: @zou3519 for associative scan.

**Comment by albertfgu** [comment]
@yanboliang I installed the nightly and it runs now!

However, I used it to profile my associative scan function and the compiled version is significantly *slower* than the uncompiled one (5x slower on a cumsum of particular reasonable shapes). I assume that's unintended behavior. I'm happy to close this issue, but let me know if the team plans to look into the speed issues (in which case I can file a separate issue) or leave it as not important for now. I know there are a lot of higher priorities :)
